File: Typescript/src/DelegationSortedIntList.ts
/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */
import { IntegerList } from './IntegerList.js'
import {SortedIntList} from "./hidden/SortedIntListLibrary";

class DelegationSortedIntList {
  private _totalAdded: number = 0
  private _sortedIntList: SortedIntList = new SortedIntList()

  constructor() {
  }

  add(num: number): boolean {
    this._totalAdded++
    return this._sortedIntList.add(num)
  }

  addAll(list: IntegerList): boolean {
    this._totalAdded += list.size()
    return this._sortedIntList.addAll(list)
  }

  get(index: number): number {
    return this._sortedIntList.get(index);
  }

  remove(num: number): boolean {
    return this._sortedIntList.remove(num);
  }

  removeAll(list: IntegerList): boolean {
    return this._sortedIntList.removeAll(list);
  }

  size(): number {
    return this._sortedIntList.size();
  }

  getTotalAdded(): number {
    return this._totalAdded;
  }
}

export { DelegationSortedIntList }

File: Java/src/main/java/edu/cmu/cs/cs214/rec04/DelegationSortedIntList.java
package edu.cmu.cs.cs214.rec04;

/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 */

public class DelegationSortedIntList implements IntegerList {
    // Keeps the total number of attempted insertions
    private int totalAdded;
    // Keeps track of the number of times elements are attempted to be added
    private SortedIntList sortedIntList;

    public DelegationSortedIntList() {
        this.sortedIntList = new SortedIntList();
    }

    /**
     * Adds the specified int to the list.
     *
     * @param num an integer to be added to the list
     * @return true if the list is changed as a result of the call
     */
    @Override
    public boolean add(int num) {
        this.totalAdded++;
        return sortedIntList.add(num);
    }

    /**
     * Adds all of the elements of the IntegerList to the list.
     *
     * @param list IntegerList containing elements to be added to the list
     * @return true if the list changed as a result of the call
     */
    @Override
    public boolean addAll(IntegerList list) {
        this.totalAdded += list.size();
        return sortedIntList.addAll(list);
    }

    /**
     * Returns the integer at the specified position in this list.
     *
     * @param index index of the element to return
     * @return the element at the specified position in this list
     */
    @Override
    public int get(int index) {
        return sortedIntList.get(index);
    }

    /**
     * Removes the first occurrence of the specified element from the list,
     * if it is present (optional operation).
     *
     * @param num an integer to be removed from the list, if present
     * @return true if an element was removed as a result of this call
     */
    @Override
    public boolean remove(int num) {
        return sortedIntList.remove(num);
    }

    /**
     * Removes from the list all of its elements that are contained in the
     * specified IntegerList.
     *
     * @param list IntegerList containing elements to be removed from
     *             the list
     * @return true if the list changed as a result of the call
     */
    @Override
    public boolean removeAll(IntegerList list) {
        return sortedIntList.removeAll(list);
    }

    /**
     * Returns the number of elements in this list. If this list contains
     * more than Integer.MAX_VALUE elements, returns Integer.MAX_VALUE.
     *
     * @return number of elements in the list
     */
    @Override
    public int size() {
        return sortedIntList.size();
    }

    /**
     * Returns the total number of attempted element insertions.
     *
     * This method returns the number of times elements have been attempted to be
     * added to the list, including duplicate insertions.
     *
     * @return the total number of attempted insertions
     */
    public int getTotalAdded() {
        return this.totalAdded;
    }
}